### DISCUSSION TITLE
Tasmota hangs after using the file system

### DISCUSSION THREAD

**Discussion by Papablaubaer** [arendst/Tasmota#21725] [Support request]
I've been working with Berry programming for some time now. It is clear that Tasmota often hangs after writing a file. This happens regardless of how the file is written, whether with the integrated file editor or via the upload function, or after uploading with Curl. I use Tasmota 14.1 on various Shellys and on the Sonoff ZigBee Bridge. The only way to revive it is to disconnect the supply voltage.

**Comment by sfromis** [comment]
I've not seen such when writing files. Did you check if you're running out of memory, like heap falling "too much"?

**Reply by Papablaubaer** [reply]
It seems to me that this has only been the case since version 14.1. How can I determine whether I'm running out of memory and, if in doubt, prevent it? Sometimes the hang occurs after a delay, then a “Restart 1” prevents this. Sometimes this happens immediately.

**Comment by arendst** [comment]
To see memory running out enable ``SO130 1`` and look at the console log timestamp now showing memory and heap usage too.

You might want to enable the ESP32 watchdog for further analysis (altough it probably restarts without useful information).

Compile your own binary with ``#define USE_ESP32_WDT`` and see if the "hang" changes in a restart due to watchdog kicking in. If that's the case then it's time for our "filesystem upload programmer" to have a look at why the watchdog kicks in probably caused by a while loop taking too long....

**Reply by Papablaubaer** [reply]
If I really think about it, the described thing only happens after Tasmota has been running for a long time. I have now made the suggested settings. Is there anything wrong with activating the ESP32 watchdog in my projects? If I understand correctly, the watchdog would ensure that the device restarts, that would definitely be a big help.

**Reply by sfromis** [reply]
The potential downside of enabling the watchdog would be that it bites when you don't think it should. This is the reason why it is not enabled by default. But if you do not get stability issues, no problem.

**Reply by Jason2866** [reply]
Are you sure it is dead? Uploading big files (> 1MB) take its time. 
Could you provide the file you are trying to load in the FS? So others can test.

**Reply by sfromis** [reply]
While I've not seen issues with Tasmota hanging, doing testing with a 17K upload (of [ftp.be](https://github.com/arendst/Tasmota/blob/development/tasmota/berry/drivers/ftp.be)), I've seen a couple of oddities. It worked on some builds (with or without PSRAM), but on another (including PSRAM) I saw it bail out with "Upload Failed - File too large". Recent development build from this week. Mentioned here too: https://github.com/arendst/Tasmota/pull/21657#issuecomment-2178909700

Doing copy/paste of the same code into "Create and edit new file", still failed, no web UI message, but `UFS: UfsEditor: write error on '/ftp.be'` in the console, and resulting file had a size of 0 bytes. On yet another device, I saw the upload work, but the confirmation page did not include the usual buttons to proceed from the page.

Makes me wonder if there is some "unstable" issue lurking, with varying "imperfect" outcomes?

**However**, upgrading to a fresh development build with newest source, trying the same test case now worked.

**Reply by Papablaubaer** [reply]
@Jason2866: This happens in all sorts of circumstances, once when I used the integrated editor to create an "autoexec.be" and wrote a line "load("fcp.be"), once when I tried to upload "PartitionWizzard.tapp" using the upload function, on other occasions when I tried to upload my Berry files using curl. All of them are basically really small files.

@sfromis: I can't really pinpoint when this happens, but it seems as if all ESP32s I use are affected, whether ESP32-D0WD-V3 v3.1, ESP32-U4WDH-D v3.1, ESP32-C3 v0.4. The devices run for months without a problem, but if I want to do something in Berry and upload it happens again and again. I've done a lot with Berry on the ZigBee Bridge in particular, once it works, it works.

**Reply by sfromis** [reply]
As hinted above, I'd suggest trying with the latest development build. While I'm not sure of details, my experience was that the other issues I mentioned seems to be better now.
